### ISSUE TITLE
Serialisation of Structs and BigEndian

### ISSUE THREAD

**Issue by mzaks** [mustiikhalil/FlatBuffer#26]
If we support BigEndian, we need to check if current endianness is BigEndian. If so we can't just serialise structs as is, we need to serialise them property by property and revert the byte order for every property.

https://github.com/mustiikhalil/FlatBuffer/blob/b02060ac72928e9da15b74ef6f4ef72d35f1f467/Sources/FlatBuffers/FlatBuffersBuilder.swift#L291

E.g.:

```
struct Test {
  let a: Int32
  let b: Float32
}

let t = Test(a:1, b:0.5)
```

The memory of object `t` in little endian is (hex):
`x01 x00 x00 x00 | x00 x00 x00 x3F`
And in BigEndian
`x00 x00 x00 x01 | x3F x00 x00 x00`

**Comment by mzaks** [comment]
Or rather, it should be fixed here:
https://github.com/mustiikhalil/FlatBuffer/blob/master/Sources/FlatBuffers/FlatBuffer.swift#L54

**Comment by mustiikhalil** [comment]
@mzaks I found out that there is an issue with serializing structs as is, yesterday I opened the [PR](https://github.com/google/flatbuffers/pull/5603) and they requested us to implement some test cases. and I started implementing them and one of the test cases was the following struct object, 
```
struct Test { a: short; b: byte }
struct Vec3 (force_align: 8) {
  x:float;
  y:float;
  z:float;
  test1:double;
  test2:Color;
  test3:Test;
}
```
the nested structs present and issue for us, and also mixing and matching the types of double, and floats in the structs will always lead to a miss aligned pointer, was thinking of switching the implementation to use `C#` way of implementing structs. 

what do you think?

**Comment by mustiikhalil** [comment]
And regarding the big endian, currently I stopped its support, you can see the FlatBufferBuilder class and also the table class

**Comment by mzaks** [comment]
I am not sure what kind of problem we exactly have.
I translated the definitions of `Color`, `Test` and `Vec3` into Swift types:
```
enum Color: UInt8 {
  case Red = 0 // color Red = (1u << 0)
  /// \brief color Green
  /// Green is bit_flag with value (1u << 1)
  case Green
  /// \brief color Blue (1u << 3)
  case Blue = 3
}

struct Test {
    let a: Int16
    let b: Int8
}
struct Vec3  {
  let x:Float32;
  let y:Float32;
  let z:Float32;
  let test1:Double;
  let test2:Color;
  let test3:Test;
}
```

And here is an example of the byte array which an instance of `Vec3` occupies in Swift:
```
var vec = Vec3(x: 0.5, y: 0.5, z: 0.5, test1: 0.1, test2: Color.Blue, test3: Test(a: 300, b: 12))
let data = Data(buffer: UnsafeBufferPointer(start: &vec, count: 1))
print(data.map{$0})
/// Result: [0, 0, 0, 63, 0, 0, 0, 63, 0, 0, 0, 63, 0, 0, 4, 0, 154, 153, 153, 153, 153, 153, 185, 63, 2, 16, 44, 1, 12, 0, 0, 0]
```

So to break it down:
```
0, 0, 0, 63 : Vector3.x
0, 0, 0, 63 : Vector3.y
0, 0, 0, 63 : Vector3.z
0, 0, 4, 0 : is 4 bytes padding don't bother about 4 on third place it is just random value
154, 153, 153, 153, 153, 153, 185, 63 : Vector3.test1
2 : Vector3.test2
16, 44 : Vector3.test3.a
1 : padding, this seems to be a problematic padding not sure if FlatBuffers struct will actually add a padding between value `a` and `b`
12 : Vector3.test3.b
0, 0, 0 : final padding
```

The struct is actually 8 bytes aligned, the padding inside of the Test struct is weird and may be breaking. Is this what is breaking in the test?

**Comment by mzaks** [comment]
Ah sorry I broke down the array wrongly, the `Test3(a:300, b:12)` by itself is represented as 

```
[44, 1, 12, 0]
```

So the padding is after Vector3.test2, which makes much more sense:
```
0, 0, 0, 63 : Vector3.x
0, 0, 0, 63 : Vector3.y
0, 0, 0, 63 : Vector3.z
0, 0, 4, 0 : is 4 bytes padding don't bother about 4 on third place it is just random value
154, 153, 153, 153, 153, 153, 185, 63 : Vector3.test1
2 : Vector3.test2
16 : padding (ignore the value)
44, 1 : Vector3.test3.a
12 : Vector3.test3.b
0, 0, 0 : final padding
```
I think the FlatBuffers structs should have similar structure with one difference the padding bytes are zeroed out. Which actually not important as they should not be read/accessed in the first place.

**Comment by mustiikhalil** [comment]
Sorry, yesterday I was typing from my phone and that's why the formatting was a bit off.

so if you would look at the PR they requested us to implement the reader for the test data they provide in their test cases suite. the issue was the following `fbs Struct` object where they have nested `Struct` objects where our code currently doesn't support. I tried using both methods, writing directly to the buffer as we do with out simple structs

```swift 
struct Vector: Writable {
     let x: Float32
     let y: Float32
     let z: Float32
}
```
writing the following struct works just by using the `_memory.storeBytes` method in `func push<T: Writeable>(struct value: T, len: Int)`. i decided to complete the struct implementation before writing the test case for the reader. since we would also need to write the same object and extract it as a file later on. The issue now is that the new struct that's written in `fbs` will always break our `storeBytes` method since there will be a `Fatal error: storeBytes to misaligned raw pointer`

```
struct Test { a: short; b: byte }
struct Vec3 (force_align: 8) {
  x:float;
  y:float;
  z:float;
  test1:double;
  test2:Color;
  test3:Test;
}
```

so describing the issue above, I tried to implement the `C#` solution where everything is basically written in place
```CSharp
builder.Prep(8, 32);
builder.Pad(2);
builder.Prep(2, 4);
builder.Pad(1);
builder.PutSbyte(test3_B);
.
.
```
but that also didn't work out well because of a `misaligned raw pointer` and today in the morning I looked at your solution and found out that there will be a `stack buffer overflow`, since I have the address sanitizer on (I started using it after the issue we had when the code just crashes in the test cases #19 ). 

```swift
 var vec = Vec3(_x: 1.0, _y: 2.0, _z: 3.0, d: 3.0, color: Color4.red, test: Test_1(_a: 5, _b: 6))
 let buffer = Data(buffer: UnsafeBufferPointer(start: &vec, count: 1))   // Stack buffer overflow
 print(Array(buffer))
```

however I found out there might be a solution to this issue, which would return the array of bytes the struct is using, without the `0, 0, 0 : final padding`

```swift
        var vec = Vec3_W(_x: 1.0, _y: 2.0, _z: 3.0, d: 3.0)  //color: Color4.red, test: Test_1(_a: 5, _b: 6))
        let values = withUnsafeBytes(of: &vec) { (buffer) -> [UInt8] in
            return Array(buffer)
        }
```
what do you think the best approach would be going forward?

**Comment by mzaks** [comment]
I am surprised about `Stack buffer overflow`, but my guess it rather has something to do with `Data`.
So from my point of view.
`MemoryLayout<T>.size` should give us a proper aligned size of the struct.
`MemoryLayout<T>.alignment` gives us the proper struct alignment at least as defined in Swift. We might get into problems if the `struct Vec3 (force_align: 8)` forces some other alignment on us.
But if we don't consider the `force_align` attribute for now. 
```
withUnsafeBytes(of: &vec) { (buffer) -> [UInt8] in
```
Gives us a buffer which we can directly copy from:
```
withUnsafeBytes(of: &vec) { (buffer)  in
    _memory.advanced(by: writerIndex - len).copyMemory(from: buffer.baseAddress!, byteCount: buffer.count)
}
```
where `len` is the `MemoryLayout<T>.size`. This way all alignments should be fine. We have a problem of garbage data in the padding though, so a unit test which does byte comparison will fail. But a unit test which does encoding / decoding roundtrip and checks the actual values, should pass just fine.

That all sad if the `force_align` attribute wants to reshape the alignment of struct, than we have to store value by value anyways, which will make the process much slower and also more complicated.

**Comment by mustiikhalil** [comment]
Yeah that should work actually, and I've already implemented it as a fix, however it will only have `garbage data` whenever we start using the padding. 
As for the `force_align` I think it wouldn't be an issue since if you read the `cpp` implementation and the `C#` one they directly use the padding, just like the implementation in swift. and their implementation in cpp is more of creating their own `struct` class that can be inlined into the memory. however I am not sure if we will run into a test cases that will require us to completely change the implementation of the structs for now. since they only test `Reading` from the buffer from different languages, and also the same language. so comparing it to our bytes won't be that necessary I guess.

you can check the implementation in #27 which is not completed yet, I will add their test cases and then request a review :D
